lib/cog/command/option_interpreter.ex

```diff
@@ -25,127 +25,123 @@ defmodule Cog.Command.OptionInterpreter do
     end
   end
 
-  def interpret(%Ast.Invocation{}=command, options, raw) do
-    case initialize(command, raw) do
-      {:ok, new_options, args} ->
-        options = Dict.merge(options, new_options, &merge_options/3)
-        {:ok, options, args}
-      error ->
-        error
-    end
-  end
+  ########################################################################
 
-  defp merge_options(_key, v1, v2) when is_list(v1) do
-    v1 ++ v2
+  defp prepare_option(%CommandOption{long_flag: lflag, short_flag: nil}=opt, acc),
+    do: Map.put(acc, lflag, opt)
+  defp prepare_option(%CommandOption{long_flag: nil, short_flag: sflag}=opt, acc),
+    do: Map.put(acc, sflag, opt)
+  defp prepare_option(%CommandOption{long_flag: lflag, short_flag: sflag}=opt, acc) do
+    acc
+    |> Map.put(lflag, opt)
+    |> Map.put(sflag, opt)
   end
-  defp merge_options(_key, v1, v2) when is_integer(v1) do
-    v1 + v2
-  end
-  defp merge_options(_key, _v1, v2) do
-    v2
-  end
-
-
-  defp set_defaults(%Ast.Invocation{}=command, options) do
-    {:ok, defs} = CommandCache.fetch_options(command, :options)
-    Enum.reduce(defs, options, &(set_default_value(&1.name, &1.option_type.name, &2)))
-  end
-
-  defp set_default_value(name, "incr", options) do
-    if Map.has_key?(options, name) == false do
-      Map.put(options, name, 0)
-    else
-      options
-    end
-  end
-  defp set_default_value(_name, _type, options) do
-    options
-  end
-
 
-  defp interpret(_defs, nil, args, state) do
-    {:ok, state, args}
-  end
-  defp interpret(_defs, [], args, state) do
-    {:ok, state, Enum.reverse(args)}
-  end
-  defp interpret(defs, [%Ast.Option{flag: flag, value: nil}=opt|t], args, state) do
-    case Map.get(defs, flag) do
+  defp interpret(_defs, nil, true_args, validated_options),
+    do: {:ok, validated_options, true_args}
+  defp interpret(_defs, [], true_args, validated_options),
+    do: {:ok, validated_options, Enum.reverse(true_args)}
+  defp interpret(defs, [%Ast.Option{name: %Ast.String{value: name}, value: nil}=opt|t], true_args, validated_options) do
+    case Map.get(defs, name) do
```

Comment:
If you're checking for the key rather than checking if it's non-nil, `Map.fetch` is probably more explicit, since it returns an ok/error tuple.